### DISCUSSION TITLE
YOLOv8 P2 and P6 variants for small object detection

### DISCUSSION THREAD

**Discussion by matijasukovic** [ultralytics/ultralytics#8227] [Q&A]
Hello, thank you for your time. I am training a model to detect small objects (as small as 10px in width) in large images (1920x1080). I saw in a similar forum post someone recommended using one of the yolov8-p2 models to improve performance on small objects. I tried them and indeed they perform better, but I still have subpar results. Looking at the models list inside of the Ultralytics library, I see there is a P6 variant. I know P2 and P6 relates to feature pyramids in some way, but I found very little information regarding P2 online, and none at all for P6. Maybe I'm looking it up wrong. Can you shed some light on what exactly is different about these models compared to the normal versions? What are the tradeoffs when using these? Will the P6 variants increase performance on small objects further?

**Comment by glenn-jocher** [comment]
Hello! I'm glad to hear you're exploring the P2 and P6 variants for small object detection. 🕵️‍♂️

The P2 and P6 variants refer to the strides of the feature pyramids used in the models. The P2 models have a stride of 2, meaning they preserve more fine-grained spatial details, which can be beneficial for detecting smaller objects. On the other hand, P6 models have a stride of 6, which makes them more suited for larger objects as they capture higher-level features.

For small object detection, P2 variants are typically more effective due to their finer feature maps. However, using P6 won't necessarily improve small object detection; it might even hinder it because of the coarser feature maps.

The tradeoffs when using these specialized models usually involve a balance between speed and accuracy. P2 models might be slower but better for small objects, while P6 models could be faster but less precise for small details.

Here's a simple example of how you might train a P2 model:

```python
from ultralytics import YOLO

# Load a P2 model
model = YOLO('yolov8n-p2.pt')

# Train the model on your dataset
model.train(data='your_dataset.yaml', epochs=50, imgsz=1920)
```

Remember to adjust the `imgsz` parameter to match your input image size for optimal results. Good luck with your training, and let us know if you have any more questions! 🚀

**Reply by park-sungho-8** [reply]
Hi @glenn-jocher.

When I was following your codes, I encountered an issue that I can't find anywhere to download 'yolov8n-p2.pt'.

Would you share a pt file or link to find it please.

**Reply by glenn-jocher** [reply]
@park-sungho-8 there are no pretrained models, but you can construct one from the provided YAML, i.e. `model=YOLO('yolov8n-p2.yaml')`

**Comment by abdulmanafpv** [comment]
Hi @glenn-jocher 
I have 4 doubts will you please help me to clarify my doubts

1) I have a doubt how we can customise the yolov8-p2.yaml for detecting more small objects for example P2/4 how we can change to P2/2 , P3/8 as P3/4 , p4/16 as P4/8, p5/32 as p5/16 so this will help me to detect the xx-small , x- small, small  and medium.

2) Also can we change any one of these for example can I change P2/4 as P2/2 without change any other  variants, if we do this like there have any problem in our model accuracy.

3) when I change the yaml file in custom way how we can use yolov8s -p2.yaml during training time. Consider if I change the yaml file name in to any other name like custom-p2.yaml so how I can get the small model architecture during training time. What would be the training command for this. ?

4)  can I change the scale factor of 2 into 4 in upsample part of head in P2 variant is there any problem when I change into 4. 

- [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 3, C2f, [256]] # 15 (P3/8-small)

**Reply by glenn-jocher** [reply]
@abdulmanafpv to customize YOLOv8 for small object detection, you can modify the feature pyramid scales in the YAML file. Adjusting layers like P2/4 to P2/2 can enhance small object detection but may affect model accuracy. You can rename the YAML file (e.g., `custom-p2.yaml`) and use it in training with `yolo train model=custom-p2.yaml`. Changing the upsample scale factor might impact performance, so test carefully. For detailed guidance, refer to the Ultralytics documentation.

**Reply by abdulmanafpv** [reply]
@glenn-jocher 

Can you help me how I can change P2/4 to P2/2 where I want to change in the yaml file?

**Reply by glenn-jocher** [reply]
@abdulmanafpv to change P2/4 to P2/2 in the YAML file, adjust the stride in the model's backbone section to reduce the downsampling rate. Make sure to test the model for performance impacts.

**Reply by nmc-eco** [reply]
@glenn-jocher Thank you for the great info!
Is it a good idea to train YOLOv8 P2 using both small and large objects simultaneously, since I'd like to detect objects with different sizes?

**Reply by glenn-jocher** [reply]
Training YOLOv8 P2 on both small and large objects can be effective, as it helps the model generalize across different object sizes. Just ensure your dataset is well-balanced.

**Reply by zeobec** [reply]
@glenn-jocher hii i have a question regarding on the model parameters. From my understanding, p2-yolov8 adds extra components compared to yolov8. Logically, the parameters will increase right? Based on my observation during training, I noticed that parameters for p2-yolov8n was much lesser than yolov8n. May i know if anyone obtained the same results like me and I wonder why that thing happened

**Reply by glenn-jocher** [reply]
@zeobec the P2 variant of YOLOv8 is designed to optimize for small object detection, which may involve architectural changes that don't necessarily increase parameter count. The reduction in parameters could be due to efficient design choices that maintain performance while using fewer resources.

**Comment by Ichiruchan** [comment]
Hello, right now I'm facing a problem. When I use YOLOv8 to detect my object, which is 20x70 (w*h), in an image sized 896x544, the results are good. However, when I try to detect an object that is 5x40 in an image sized 1024x600, the results are poor. I also tried using yolov8x-pose-p2, but it didn't meet my needs.

When I increase the bounding box size to 8x40, the results improve, but the extra part of the bounding box contains background, which doesn't provide any additional information. I believe this is because the object is too small.

If I want to detect objects smaller than what P2 can handle, what should I do? Should I change the model's YAML configuration?

**Reply by glenn-jocher** [reply]
For detecting very small objects, consider using image tiling (splitting the image into smaller patches before inference) to effectively increase the model's resolution for detection. Additionally, you can modify the model's YAML file to use a higher-resolution input size or adjust the depth and width factors for finer feature learning. Also, training the model with data augmentations such as MixUp and Mosaic can help. For further guidance, check https://docs.ultralytics.com/guides/model-evaluation-insights/#image-tiling-for-small-objects.

**Comment by ahmedbenakouche** [comment]
Hello @glenn-jocher 

I have a few questions about small object detection. I fine-tuned YOLO26n on my dataset, but I got relatively low performance on small objects (Recall = 0.272).

1. If I understand correctly, this may be partly because the default YOLO detection head uses strides 8/16/32 (P3, P4, P5).
In my case, some objects are only about 5.5 pixels wide at imgsz=640. With the deeper layers, especially with large strides, they become extremely small. For example, with stride 32, this becomes about 5.5 / 32 = 0.17 cell on the feature map, which seems very hard to use.
Is this reasoning correct?

I also increased imgsz to 1280 and got much better results compared to imgsz=640 (Recall = 0.665), but inference becomes about 2× slower.
I also tested SAHI with 9 slices and got Recall = 0.629, but it was much slower than simply increasing imgsz to 1280.

2. So my question is: in which cases is SAHI really useful?
In my case, increasing image size seems more useful than SAHI.
My guess is : if the training dataset already contains many small annotated objects, then it may be better to train with the largest possible imgsz, while SAHI is more useful when objects at inference time are much smaller than the objects seen during training.
Is that correct?

3. I also recently discovered the YOLO26-P2 variant, which may improve small object detection.
Would it be worth testing? And does it usually have a big impact on inference speed? In your opinion, is it generally better to: increase imgsz, or use the P2 variant?

4. If I use YOLO26-P2, which adds the P2 detection layer, is this new branch randomly initialized, or is it also pretrained on COCO?
Would I need to train from scratch, or is normal fine-tuning enough?
More generally, when do you recommend training from scratch, especially when the application domain is very different from COCO?

**Reply by glenn-jocher** [reply]
Suggested reply:

@AHMEDBENAKOUCHE @matijasukovic your reasoning is correct overall: small targets can disappear as stride grows, so the standard YOLO26 detect head (`P3/8`, `P4/16`, `P5/32`) is tougher on ~5–10 px objects than `yolo26-p2`, which adds a higher-resolution `P2/4` head; by contrast, the YOLOv8 `P6` configs add a coarser `P6/64` level, which is usually more about larger objects/context than tiny targets, and both variants generally cost some extra latency/memory. See the [FPN overview](https://www.ultralytics.com/glossary/feature-pyramid-network-fpn) and this [small-object guide](https://www.ultralytics.com/blog/how-to-improve-model-map-on-small-objects-a-quick-guide). ([github.com](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/26/yolo26.yaml))

If increasing `imgsz` already gives you the best recall/speed tradeoff, that’s perfectly reasonable; SAHI tends to help most when you must keep the base input smaller but still need sliced inference on large images where objects are tiny, and it is usually slower than a single forward pass at a larger `imgsz`. ([ultralytics.com](https://www.ultralytics.com/blog/how-to-improve-model-map-on-small-objects-a-quick-guide))

For `-p2`, I’d start with normal fine-tuning rather than scratch training: official YOLO26 detect weights are pretrained on COCO, and if you initialize a `-p2` architecture from a standard checkpoint the extra P2 branch has no 1:1 counterpart in the base head, so that part just learns during fine-tuning; I’d only jump to scratch if you have a lot of domain-specific data and pretraining clearly is not transferring well. ([github.com](https://github.com/ultralytics/ultralytics))

**Reply by 14790897** [reply]
I think SAHI is mainly used when the targets are much smaller than the ones in the training images. By slicing and enlarging the images, it keeps the target size consistent with the training set. I believe SAHI is only really useful in these specific cases.

**Reply by glenn-jocher** [reply]
Yes, that’s generally right: SAHI helps most when full-frame inference makes targets much smaller than the scale seen during training, or when memory limits prevent using a larger `imgsz`; if a single pass at higher `imgsz` already gives the best recall/latency tradeoff, that’s usually the better option.

**Comment by Shravasti221** [comment]
I had a quick question related to this as well. Will having a lot of sub 4px boxes affect my training in the sense that the loss functions get polluted? This is my current bounding box size distribution
<img width="425" height="370" alt="image" src="https://github.com/user-attachments/assets/8c845781-752a-4d85-94e6-c5be692c6da6" />

I am training at 1920 x 1080 resolution
Would you recommend filtering out the sub 4px bounding boxes for better results? I am using yolo26n-p2.

**Reply by glenn-jocher** [reply]
Valid sub-4 px boxes do not inherently pollute the loss, and YOLO26’s small-target-aware assignment helps preserve supervision for tiny objects; however, 1 px annotation error is substantial at that scale. Don’t filter them solely by size if they matter at deployment—prefer tiling/cropping so objects occupy more pixels, as described in [image tiling for small objects](https://docs.ultralytics.com/guides/model-evaluation-insights#image-tiling-for-small-objects), and compare validation recall with and without filtering. Remove only boxes that are ambiguous, inconsistently labeled, or below your application’s useful detection limit.

**Reply by Shravasti221** [reply]
Thank you for your inputs!

**Reply by glenn-jocher** [reply]
The standard model predicts at P3/8–P5/32, while [`yolov8-p2`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/v8/yolov8-p2.yaml) adds a stride-4 P2 head that preserves detail for tiny objects at higher memory/compute cost. [`yolov8-p6`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/v8/yolov8-p6.yaml) instead predicts at P3/8–P6/64, adding a coarser scale for very large objects and generally will not improve 10 px targets. For your case, prefer P2 with higher-resolution crops or tiling, then confirm the trade-off using size-specific validation metrics.
